internal/service/alarms/internal/dictionary_definition/dictionary_definition.go

```diff
@@ -0,0 +1,460 @@
+	}
+
+	slog.Info("Successfully processed nodeClusterType", "nodeClusterType", nodeClusterType.NodeClusterTypeId)
+	return nil
+}
+
+// upsertAlarmDictionary process dict for nodeClusterType
+func (r *AlarmDictionaryDefinition) upsertAlarmDictionary(ctx context.Context, nodeClusterType clusterserver.NodeClusterType) (models.AlarmDictionary, error) {
+	extensions, err := getVendorExtensions(nodeClusterType)
+	if err != nil {
+		return models.AlarmDictionary{}, fmt.Errorf("failed to get extensions for nodeClusterType  %s: %w", nodeClusterType.NodeClusterTypeId, err)
+	}
+	// Alarm Dictionary record
+	alarmDict := models.AlarmDictionary{
+		AlarmDictionaryVersion: extensions.version,
+		EntityType:             fmt.Sprintf("%s-%s", extensions.model, extensions.version),
+		Vendor:                 extensions.version,
+		ObjectTypeID:           nodeClusterType.NodeClusterTypeId,
+	}
+
+	// Upsert Alarm Dictionary
+	alarmDictRecords, err := r.AlarmsRepository.UpsertAlarmDictionary(ctx, alarmDict)
+	if err != nil {
+		return models.AlarmDictionary{}, fmt.Errorf("failed to upsert alarm dictionary_definition: %w", err)
+	}
+
+	if len(alarmDictRecords) != 1 {
+		return models.AlarmDictionary{}, fmt.Errorf("unexpected number of Alarm Dictionary records, expected 1, got %d", len(alarmDictRecords))
+	}
+
+	slog.Debug("Alarm dictionary upserted", "nodeClusterType", nodeClusterType.NodeClusterTypeId, "alarmDictionaryID", alarmDictRecords[0].AlarmDictionaryID)
+	return alarmDictRecords[0], nil
+}
+
+// upsertAlarmDictionary process def for nodeClusterType
+func (r *AlarmDictionaryDefinition) processAlarmDefinitions(ctx context.Context, nodeClusterType clusterserver.NodeClusterType, ad models.AlarmDictionary) error {
+	// Get filtered rules
+	filteredRules := r.getFilteredRules(nodeClusterType.NodeClusterTypeId)
+
+	// Create alarm definitions
+	records := r.createAlarmDefinitions(filteredRules, ad, nodeClusterType)
+
+	// Upsert and cleanup
+	if err := r.upsertAndCleanupDefinitions(ctx, records, nodeClusterType.NodeClusterTypeId); err != nil {
+		return fmt.Errorf("failed to upsert and cleanup definitions: %w", err)
+	}
+
+	slog.Info("Successfully processed all AlarmDefinitions", "nodeClusterType", nodeClusterType.NodeClusterTypeId, "alarmDict", ad.AlarmDictionaryID)
+	return nil
+}
+
+// getFilteredRules check to see if rule can potentially be skipped
+func (r *AlarmDictionaryDefinition) getFilteredRules(nodeClusterTypeID uuid.UUID) []monitoringv1.Rule {
+	// Upsert will complain if there are rules with the same Alert and Severity
+	// We need to filter them out. First occurrence wins.
+	type uniqueAlarm struct {
+		Alert    string
+		Severity string
+	}
+
+	var filteredRules []monitoringv1.Rule
+	exist := make(map[uniqueAlarm]bool)
+
+	for _, rule := range r.RulesMap[nodeClusterTypeID] {
+		severity, ok := rule.Labels["severity"]
+		if !ok {
+			slog.Warn("rule missing severity label", "alert", rule.Alert, "nodeClusterTypeID", nodeClusterTypeID)
```

Comment:
we shouldn't restrict this. Our application supports alerts without this as well.

Reply:
removed continue